### PR TITLE
gh-117431: Adapt str.{start,end}swith to use the METH_FASTCALL calling convention

### PR DESCRIPTION
**Pull request by erlend-aasland**
* Issue: gh-117431

### CODE REVIEW

**Comment by erlend-aasland** [comment]
cc. https://github.com/faster-cpython/ideas/issues/671

**Comment by erlend-aasland** [comment]
```
# main
$ ./python.exe -m timeit -s "s = 'abcdef'" "s.startswith('abc')"
5000000 loops, best of 5: 89.1 nsec per loop

# this PR
$ ./python.exe -m timeit -s "s = 'abcdef'" "s.startswith('abc')"
10000000 loops, best of 5: 26.4 nsec per loop
```

File: Objects/clinic/unicodeobject.c.h
Comment (eendebakpt) [inline Objects/clinic/unicodeobject.c.h:1451]: This duplicates code from `startswith`. If we also convert `count`, `find`, `rfind`, etc. and the corresponding `bytes` methods we will have more code duplication. Can we do something similar to the previous implementation to avoid this?

Comment (erlend-aasland) [reply Objects/clinic/unicodeobject.c.h:1451]: This is clinic code; it is ok with code duplication in generated code. The point[^1] with Argument Clinic is to reduce boiler plate code in the source file and put it into the generated file.

[^1]: well, one of the points of AC

**Comment by erlend-aasland** [comment]
We'll also want to convert `str.find` and friend and finally get rid of the parsing helper in strlib. However, that is best left to a follow-up PR. I prefer small and incremental changes.

File: Objects/unicodeobject.c
Comment (JelleZijlstra) [inline Objects/unicodeobject.c]: Does this text show up in the user-visible docstring? PY_SSIZE_T_MAX isn't very useful to Python users; the effective default is just "look to the end of the string". I would omit this part:
```suggestion
        Optional stop position.
```

Comment (erlend-aasland) [reply Objects/unicodeobject.c]: Yep, param docstrings are embedded by Argument Clinic into the "body" of the docstring.

Perhaps:

```suggestion
        Optional stop position. Default: the end of the string.
```

Comment (erlend-aasland) [reply Objects/unicodeobject.c]: BTW; you can see the generated output in `Objects/clinic/unicodeobject.c.h`.

Comment (JelleZijlstra) [reply Objects/unicodeobject.c]: Your suggestion sounds good!

Ah right, should have looked at the generated output.

Comment (erlend-aasland) [reply Objects/unicodeobject.c]: Thanks for pointing this out; I think the docstring looks better now!

**Review by JelleZijlstra — APPROVED** [review]

**Comment by erlend-aasland** [comment]
Thanks for the review, @JelleZijlstra and @eendebakpt. I'll hold off merging for a day or two.

**Review by methane — APPROVED** [review]

**Comment by erlend-aasland** [comment]
Well, with two core devs approving, I'll go ahead and merge.

File: Objects/unicodeobject.c
Comment (methane) [inline Objects/unicodeobject.c:13081]: I had overlooked that this should be `suffix`.

Comment (erlend-aasland) [reply Objects/unicodeobject.c:13081]: Good catch; `endswith` uses a different name. I'll fix it right away. Thanks!

Comment (erlend-aasland) [reply Objects/unicodeobject.c:13081]: - #117499

**Comment by vstinner** [comment]
Nice optimization!

**Pull request merged**
